src/http_connection.cpp

```diff
@@ -449,12 +449,14 @@ void http_connection::on_timeout(boost::weak_ptr<http_connection> p
 			error_code ec;
 			c->m_sock.close(ec);
 			if (!c->m_connecting) c->connect();
+			c->m_last_receive = now;
+			c->m_start_time = c->m_last_receive;
 		}
 		else
 		{
 			c->callback(boost::asio::error::timed_out);
+			return;
 		}
-		return;
```

Comment:
yes, but `m_start_time` isn't reset until `on_connected()` is called.

in this test case `m_connecting` is always `false` (the timeout happens after a successful connection), but  we probably need another test covering the case where the connect never succeeds and just hangs in the air (often the case for firewalled connections where packets are dropped).